lms/djangoapps/grades/course_grades.py

```diff
@@ -0,0 +1,256 @@
+"""
+Functionality for course-level grades.
+"""
+from logging import getLogger
+from django.conf import settings
+import dogstats_wrapper as dog_stats_api
+import random
+
+from opaque_keys.edx.keys import CourseKey
+from openedx.core.djangoapps.signals.signals import GRADES_UPDATED
+
+from course_blocks.api import get_course_blocks
+from courseware.courses import get_course_by_id
+from courseware.model_data import ScoresClient
+from student.models import anonymous_id_for_user
+from util.db import outer_atomic
+from xmodule import graders, block_metadata_utils
+from xmodule.graders import Score
+
+from .context import grading_context
+from .scores import get_score
+
+
+log = getLogger(__name__)
+
+
+def iterate_grades_for(course_or_id, students, keep_raw_scores=False):
+    """Given a course_id and an iterable of students (User), yield a tuple of:
+
+    (student, gradeset, err_msg) for every student enrolled in the course.
+
+    If an error occurred, gradeset will be an empty dict and err_msg will be an
+    exception message. If there was no error, err_msg is an empty string.
+
+    The gradeset is a dictionary with the following fields:
+
+    - grade : A final letter grade.
+    - percent : The final percent for the class (rounded up).
+        - section_breakdown : A breakdown of each section that makes
+        up the grade. (For display)
+    - grade_breakdown : A breakdown of the major components that
+        make up the final grade. (For display)
+    - raw_scores: contains scores for every graded module
+    """
+    if isinstance(course_or_id, (basestring, CourseKey)):
+        course = get_course_by_id(course_or_id)
+    else:
+        course = course_or_id
+
+    for student in students:
+        with dog_stats_api.timer('lms.grades.iterate_grades_for', tags=[u'action:{}'.format(course.id)]):
+            try:
+                gradeset = summary(student, course, keep_raw_scores)
+                yield student, gradeset, ""
+            except Exception as exc:  # pylint: disable=broad-except
+                # Keep marching on even if this student couldn't be graded for
+                # some reason, but log it for future reference.
+                log.exception(
+                    'Cannot grade student %s (%s) in course %s because of exception: %s',
+                    student.username,
+                    student.id,
+                    course.id,
+                    exc.message
+                )
+                yield student, {}, exc.message
+
+
+def summary(student, course, keep_raw_scores=False, course_structure=None):
```

Comment:
I like the name 'summary' here. It seems more faithful to the purpose of this function than `grade`.